### PR TITLE
GEODE-3940: fix deadlock in backup messages

### PR DESCRIPTION
**Pull request by kirklund**
There are 3 separate commits in this PR. 1) reintroduce Nick's commit for GEODE-3801, 2) Lynn's fix for deadlock in backup messages, 3) prevent logger warnings when sender processes the backup messages.

commit d55e6d6cae348fedb9f1341a9abde2096be3bead (HEAD -> feature/GEODE-3940, origin/feature/GEODE-3940, kirklund-fork/feature/GEODE-3940)
Author: Kirk Lund <klund@apache.org>
Date:   Wed Nov 8 10:25:07 2017 -0800

    GEODE-3940: prevent warning when processing message by sender
    
    Prevents logging warning when processing these message types by sender:
    
    * FinishBackupRequest, FlushToDiskRequest, PrepareBackupRequest

commit 81ff81da9fc121607b35aca15d99e24cf0783e7f
Author: Lynn Gallinat <lgallinat@pivotal.io>
Date:   Fri Nov 3 17:09:42 2017 -0700

    GEODE-3940: fix deadlock in backup messages
    
    * introduce unit tests for backup messages
    * add unit test for BackupLock

commit 080730b2df5c7ec0a17acd195a5b953dfc6a0d5c
Author: Nick Reich <nreich@pivotal.io>
Date:   Tue Oct 24 08:24:42 2017 -0700

    GEODE-3801: Use hardlinks for backup oplog files (#963)
    
    * Oplog files that are backed up are read-only. For this reason,
      they can be transfered to the backup location through hard links,
      instead of copying the file. This change improves speed of backups.
      If the creation of a hard link fails, we revert to the existing copy
      behavior.
    * During backups, the copying of the oplog's krf file was being done while
      that file could still be in the process of writing. This change ensures
      that if a krf is to be written, that it is finished and included in the
      backup
    * cleanup existing oplog tests

### CODE REVIEW

File: geode-core/src/test/java/org/apache/geode/admin/internal/FinishBackupRequestTest.java [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+    expectedResults.put(localMember, localMember_PersistentIdSet);
+    expectedResults.put(member1, member1_PersistentIdSet);
+    expectedResults.put(member2, member2_PersistentIdSet);
+    when(replyProcessor.getResults()).thenReturn(expectedResults);
+
+    Map<DistributedMember, Set<PersistentID>> results = finishBackupRequest.send();
+
+    assertThat(results).isEqualTo(expectedResults);
+  }
+
```
Comment (nreich) [inline geode-core/src/test/java/org/apache/geode/admin/internal/FinishBackupRequestTest.java]: Above comments apply to below test in this class and to similar tests in other classes.

File: geode-core/src/test/java/org/apache/geode/admin/internal/FinishBackupRequestTest.java [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+  @Test
+  public void getRecipientsDoesNotIncludeNull() throws Exception {
+    InternalDistributedMember nullMember = null;
+
+    assertThat(finishBackupRequest.getRecipients()).doesNotContain(nullMember);
+  }
+
+  @Test
+  public void sendShouldUseDMToSendMessage() throws Exception {
```
[the remark this answers is not in the document]
Comment (kirklund) [reply geode-core/src/test/java/org/apache/geode/admin/internal/FinishBackupRequestTest.java]: I'm actually following interaction-based testing on purpose. For interaction-based testing (vs state-based testing), you perform white-box testing by verifying that certain methods were invoked on collaborators. This is supposedly the most common approach in unit tests that use mocking. Without doing this, there's not much testing you could really do in a unit test. If these message classes had been written using TDD, these are also the sort of tests you would end up with. Let me know if you want to talk about this more.

File: geode-core/src/main/java/org/apache/geode/admin/internal/FinishBackupRequest.java [hunk 3]
```diff
@@ -48,68 +47,92 @@
 /**
  * A request send from an admin VM to all of the peers to indicate that that should complete the
  * backup operation.
- *
- *
  */
 public class FinishBackupRequest extends CliLegacyMessage {
   private static final Logger logger = LogService.getLogger();
 
+  private final DM dm;
```
Comment (dschneider-pivotal) [inline geode-core/src/main/java/org/apache/geode/admin/internal/FinishBackupRequest.java]: Lets try to refactor the send logic and reply logic into another class whose job it is to create this message, send it and then wait for the reply.

File: geode-core/src/test/java/org/apache/geode/admin/internal/FinishBackupRequestTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (nreich) [reply geode-core/src/test/java/org/apache/geode/admin/internal/FinishBackupRequestTest.java]: Thanks for explaining the testing philosophy. I can see the logic of trying to handle these interactions on a unit test level.

**Comment by kirklund** [comment]
Closing this PR as the code is now stale. We'll submit a new PR sometime this week.

**Pull request closed without merging**